client/server/render/index.js

```diff
@@ -84,11 +83,19 @@ function render( element, key, req ) {
 
 		let renderedLayout = markupCache.get( key );
 
-		logServerEvent( req.context.sectionName, {
-			// Note: "ssr" just categorizes the stat. It doesn't necessarily mean SSR was used for the request.
-			name: `ssr.markup_cache.${ renderedLayout ? 'hit' : 'miss' }`,
-			type: 'counting',
-		} );
+		/**
+		 * If the SSR pipeline executed thinking that a layout would exist, it
+		 * likely skipped several important steps needed to generate a layout,
+		 * such as fetching the data rendered on the page. As a result, we can't
+		 * confidently render a page and the default error handler will allow
+		 * the client to render it instead.
+		 *
+		 * This should not happen -- if it does, we need to investigate and fix it.
+		 */
+		if ( req.context.hasLayout && ! renderedLayout ) {
```

Comment:
This condition can actually happen in practice, and it doesn't need to be a bug in code. Values in the cache can get evicted when there are more than 3000 keys of if they are older than 1 hour. And our route handler looks at the cache at two places, possibly separated in time:
```js
if ( markupCache.get( key ) ) {
  context.hasLayout = true;
}

await ( /* intermediate handlers calling next() asynchronously */ )

let renderedLayout = markupCache.get( key );
```
The `key` may no longer be in the cache on the second call.

One solution is to store the actual markup to the `context`, not only a boolean:
```js
context.cachedLayout = markupCache.get( key );
```

There are various edge cases where both `markupCache.get( key )` calls can return different results -- one of them being `null`, both being non-null but different... Because keys can be evicted and other interleaved requests can write new values. They don't seem dangerous, only the case where both calls return a different non-null value is weird. That would mean that the requests are not deterministic -- rendering different markup on each request.

Reply:
Great idea, I like this a lot more.

> That would mean that the requests are not deterministic -- rendering different markup on each request.

Would that mean the LRU code is buggy?

Reply:
Refactored to do this 👍 (I also changed the name to `cachedMarkup` instead of `cachedLayout`, because in the `render` function, the given "layout" is rendered into markup which is added to the cache)